src/main/scala/util/RecordMap.scala

```diff
@@ -0,0 +1,34 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.util
+
+import chisel3._
+import scala.collection.immutable.ListMap
+import chisel3.internal.requireIsChiselType
+import chisel3.experimental.DataMirror.internal.chiselTypeClone
+
+final class RecordMap[T <: Data](eltMap: ListMap[String, T])
+    extends Record {
+  
+  eltMap.foreach { case (name, elt) => requireIsChiselType(elt, name) }
+
+  def apply(x: Int) = eltMap.values.toSeq(x)
+  def apply(x: String) = eltMap.get(x)
+  def size = eltMap.size
+
+  // This is needed for Record, and doesn't give the actual elements
+  val elements = ListMap[String, T]() ++ eltMap.mapValues(chiselTypeClone(_).asInstanceOf[T])  // mapValues return value is lazy
+  def data = elements.values
+
+  override def cloneType: this.type = (new RecordMap(eltMap)).asInstanceOf[this.type]
+
+}
+
+object RecordMap {
+
+  def apply[T <: Data](eltMap: ListMap[String, T]) = new RecordMap(eltMap)
+
+  def apply[T <: Data](elements: (String, T)*) {
```

Comment:
```suggestion
  def apply[T <: Data](elements: (String, T)*): RecordMap[T] =  {
```

oh are you pointing out that I am just missing an `=` and this is a pointless, untested bit of code... yes.